Add a packet.PrivateKey.Encrypt() method

dvusboy opened the pull request:
This is the counterpart to the packet.PrivateKey.Decrypt() method.
The intention is to produce a password-protected PrivateKey that
can be serialized.

* S2K Usage is always 254 (RFC-4880 5.5.3)
* S2K Specifier Type is always Iterated and Salted
  (RFC-4880 3.7.1.3)

kcwong-verseon committed:
Add a packet.PrivateKey.Encrypt() method

This is the counterpart to the packet.PrivateKey.Decrypt() method.
The intention is to produce a password-protected PrivateKey that
can be serialized.

* S2K Usage is always 254 (RFC-4880 5.5.3)
* S2K Specifier Type is always Iterated and Salted
  (RFC-4880 3.7.1.3)

dvusboy commented:
I'll work on fixing the failed tests.

dvusboy commented:
There's a silly mistake, so I revised the commit. The point of the PR is so that I can encrypt the private/secret keys and eventually export the entities all from within this library.

maxtaco commented:
@zapu can you do the code review?

zapu commented:
On it! Thank you for the PR.

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -155,6 +156,56 @@ func mod64kHash(d []byte) uint16 {
+}
+
 func (pk *PrivateKey) Serialize(w io.Writer) (err error) {
 	// TODO(agl): support encrypted private keys
```
Is it still `TODO`? :)

dvusboy replied:
I guess not. Do you think the implementation is sound? I'm very much a Golang noob.

dvusboy replied:
Maybe we should ping agl?

maxtaco replied:
@agl hasn't been hacking on this library for a bit, so my guess is that it's still TODO.

zapu replied:
I just thought that since this PR introduces "encrypted private keys", maybe this comment is no longer true. But we can keep it :)

dvusboy replied:
I'll delete it then.

zapu commented on openpgp/packet/private_key_test.go:
```diff
@@ -56,6 +65,77 @@ func TestPrivateKeyRead(t *testing.T) {
+			continue
+		}
+
+		switch pubKey := privKey.PublicKey.PublicKey.(type) {
```
There is easier way of doing roundtrip encrypt-decrypt in tests, see here for example: https://github.com/keybase/go-crypto/blob/master/openpgp/ecdh_test.go#L121-L141 Unless I misunderstood the intent.

dvusboy replied:
That is the intent: That I can use the original key (public-half, obviously) to encrypt a message, and then I can use the copied version, which went through a new passphrase, serialization, and parsing, to decrypt the message. I'll see what I can do.

zapu replied:
I just tried to make it `privKey.Decrypt` after encrypting the message (so priv was still encrypted while it used pub to encrypt message) and it worked. Cool!

Maybe we could make it two tests, one with simple `priv.Decrypt; priv.Encrypt; priv.Decrypt` and some assertions that it worked, and another one with that roundtrip message encryption. What do you think?

dvusboy replied:
If I adopt the `ecdh_test.go` approach, I'd have to import `openpgp` and I ended up with an import cycle. So, the compile failed. Any suggestion?

zapu replied:
I'm, right, so a test doing round-trip key encryption and round-trip message encryption should be in some other file in openpgp package. @maxtaco what do you think?

dvusboy replied:
Since the 2 changes I've made primarily affect `PrivateKey` encryption and serialization. Perhaps I should limit to testing those 2 things and leave the message encrypt-decrypt roundtrip to other tests. What do you think?

zapu replied:
You're probably right, we are mostly concerned that it can be serialized and deserialized when encrypted, and if it can be decrypted. There are other tests responsible for message encryption. However I like the idea of trying to decrypt a message to see if we end up with functional key after passphrase decryption.

dvusboy replied:
Let's see if I can think of something in a 3rd commit.

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -174,27 +225,27 @@ func (pk *PrivateKey) Serialize(w io.Writer) (err error) {
+		_, err = buf.Write([]byte{
+			254,             // SHA-1 Convention
+			byte(pk.cipher), // Encryption scheme
+		})
```
Can we make it like that?
```
	s2kConfig := s2k.Config{
		Hash: config.Hash(),
		S2KCount: 0,
	}
```

looks more readable for me

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -155,6 +156,56 @@ func mod64kHash(d []byte) uint16 {
+
+	pk.sha1Checksum = true
+	pk.cipher = config.Cipher()
+	s2kConfig := s2k.Config{
```
Can we make it:
```
	s2kConfig := s2k.Config{
		Hash: config.Hash(),
		S2KCount: 0,
	}
```

looks more readable for me

dvusboy replied:
right, you are.

zapu commented:
I've checked against `gpg2` if I could import a key generated and encrypted by this PR, and it seemed to work. Thank you again for this PR! I will do another read through the code tomorrow.

kcwong-verseon committed:
Address code review concerns from zapu

* readability improvement
* dropped the complicated message roundtripping test and limit to
  just validating the changes

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -155,6 +156,56 @@ func mod64kHash(d []byte) uint16 {
 }
 
+// This is the counterpart to the Decrypt() method below
+// @param passphase is the passphrase used to generate the derived-key based
```
Generally we don't have doxygen-style comments in go (at least not in go-crypto). Could you try to rewrite it to something like this: https://github.com/keybase/go-crypto/blob/master/openpgp/write.go#L204-L208 ? It also describes that the arguments and return values but in more prose-like way. Then we will have everything consistent. Thank you!

kcwong-verseon committed:
Drop the doxygen-style comment on the parameters to Encrypt()

* spells out a bit more on the hardcoded defaults: S2K count of
  0, S2K usage of 254. The S2K specifier type of Iterated + Salted,
  3, is set by s2k.Serialize().

kcwong-verseon committed:
Add a round-trip message encryption/decryption test.

* leverage an existing case for TestSignedEncryptedMessage()
* load the keyring, find the one used for encryption
* change the passphrase on all encrypted private keys in that
  keyring
* change the prompt function to decrypt the private key with
  the new passphrase
* verify the message decrypt successfully and to expected value

dvusboy commented:
Added a new test case in `openpgp/read_test.go` to do the passphrase change and then decrypt the message from a previous test case.

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -155,6 +156,60 @@ func mod64kHash(d []byte) uint16 {
 	return h
 }
 
+// This is the counterpart to the Decrypt() method below.
```
`private_key.go:159:1: comment on exported method PrivateKey.Encrypt should be of the form "Encrypt ..."`

one from `golint`. Generally the comment should be phrased like: `// Encrypt is the counterpart to the Decrypt`, that's the convention.

Also please run `go fmt private_key.go`, it will reformat the code and save it automatically. It fixes one line in this file, rest are fine IIRC.

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -174,27 +229,27 @@ func (pk *PrivateKey) Serialize(w io.Writer) (err error) {
+			254,             // SHA-1 Convention
+			byte(pk.cipher), // Encryption scheme
+		})
+		if err != nil {
```
I'm wondering why are we checking for errors here, when `buf` is `bytes.Buffer`. I even checked https://golang.org/src/bytes/buffer.go and it does not seem that Write operations on Buffer ever end up with an error. @maxtaco am I missing something? Or is it a good convention to always check for errors?

dvusboy replied:
Like I said at the beginning, I'm a Golang noob. So, if a method returns `error`, I'm going to check it. And you are correct that the current `Buffer.Write()` only returns `nil` for `error`. Is there any harm in checking?

kcwong-verseon committed:
Additional review comments

* update comment on Encrypt() to be more idiomatic
* gofmt cleanup
* remove agl's TODO comment in Serialize()

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -217,7 +276,7 @@ func (pk *PrivateKey) Serialize(w io.Writer) (err error) {
 	}
 
-	if len(privateKeyBytes) > 0 {
+	if len(privateKeyBytes) > 0 && !pk.Encrypted {
```
This branch is not taken when `pk.Encrypted` because the checksum is already in `privateKeyBytes` (in encrypted form), right?

dvusboy replied:
Correct.

dvusboy commented:
BTW, I ran into an unrelated problem.
If one calls `openpgp.NewEntity()` and create a new `Entity`, you cannot immediately call `entity.Serialize()` because none of the signatures are actually created. You'd have to call [`entity.SerializePrivate()`](https://github.com/keybase/go-crypto/blob/master/openpgp/keys.go#L709-L714) first and then you can call `entity.Serialize()`. Why isn't the signing done during construction? Do you know? I think this is @agl legacy code.

zapu commented:
But `entity.SerializePrivate()` returns serialized entity with private key and signatures, there's no reason to call `Serialize()` after.

I have this code for some keygen testing: https://gist.github.com/zapu/e58128dc1a691f3a1c18f61c6e6170e4

dvusboy commented:
The reason is that I only need the public data to be sent to another service for data encryption. I just find it surprising that `NewEntity()` does not provide a object that can be readily serialized. It is not intuitive at all, that, in order to generate the serialized public data, you need the side-effect of `SerializePrivate()`.  In that respect, I consider it a bug. It is tangential to the objective of this PR. Thank you for all your help though, @zapu.

zapu commented:
Hm, it sounds like a valid workflow to be able to serialize not-signed keys or subkeys, even though they would not be usable in any pgp program. But in general we try not to change any go-crypto APIs, just improve upon existing code.

About this PR, I'm testing few more things, but Max will need to give a final :+1: to this, as he is the crypto expert :) Thank you for your patience.

my only issue right now is that `Decrypt` does not clear the `Encrypted` flag, but on the other hand I can understand this logic. In order to decrypt and serialize not-protected key, one needs to do
```
entity.PrivateKey.Decrypt([]byte("hello"))
entity.PrivateKey.Encrypted = false
```

So it's up to discussion how the API should work - for example if `Encrypt` should clear or zero out "clear-text" private key data or keep it, or if Decrypt should mutate the `Encrypted` field. Again, let's stick to what golang/crypto does.

zapu commented:
Another thing is that once you import encrypted key, you `Decrypt()` it and do not manually clear `Encrypted` flag, it will fail to serialize. Also importing encrypted key and serializing it back is something that I would expect to work (and just serialize encrypted data back) as a consumer. So maybe parsing encrypted keys should fill the `s2kHeader` buffer properly, I don't know if that's possible. Or at least make exporting fail when `s2kHeader` buffer is but key says it's encrypted.

dvusboy commented:
@zapu, I'm not sure about having `Decrypt()` clear the `Encrypted` flag. I think the issue rest solely in `s2kHeader`. I'll see what I can do.

dvusboy commented:
@zapu I did not realize `cfb.XORKeyStream()` would clear the source buffer, which means after `Decrypt()` is called. The `encryptedData` field would be empty. So, I can:

1. Unset `Encrypted` like you said, and unless `Encrypt()` is called, the next time the private key is serialized, it will not be encrypted. Or,
2. Save the encrypted form in `Decrypt()` so that you would still serialize back into the encrypted form without first calling `Encrypt()`.

I'm not certain which approach I should take.

kcwong-verseon committed:
Add a test for serialization after decrypt

* this is to make sure PrivateKey.s2kHeader does not trip up the
  serialization

dvusboy commented:
@zapu I've been exploring this some more. So, here's what I've found:

`Decrypt()` *does* clear both `encryptedData` and `Encrypted` flag, just not directly. It is done in all those `parse*PrivateKey()` methods that are specific to the key type ([see example](https://github.com/keybase/go-crypto/blob/master/openpgp/packet/private_key.go#L366-L367)). This means after `Decrypt()` is called, the private key is no longer encrypted. I've added a test to validate the added field, `s2kHeader`, doesn't trip up serialization after decrypt.

As for the API, I'm not sure it makes sense that the private key cannot be used after `Encrypt()` is called. Since this is completely new, we are definitely in uncharted territory. Right now what we have is:

> If the serialized form is of an encrypted private key, just calling `Read()` will not result in a usable `PrivateKey`, you must call `Decrypt()` first. This is similar to what happens when you do a `gpg --import`: you'll be prompted for the passphrase.
> But once `Decrypt()` is called, the `PrivateKey` is *no longer encrypted*. It can be used, but exporting it will result in an unencrypted payload.
> By adding `Encrypt()`, it is now possible to protect a bare `PrivateKey`. Serializing this `PrivateKey` will also produce an encrypted export.

The open question is: Should you be able to use an encrypted `PrivateKey`? I think the design of the original API is leaning towards no. Instead, to produce an encrypted export, `Serialize()` needs to ask for a passphrase. This is, BTW, a very different behavior than, say `gpg`. I'm wondering if I should change `Decrypt()` to keep the encrypted state around to make it all consistent with `Encrypt()` leaving the bare private key around. Changing the signature to `Serialize()` is not appetizing to me.

zapu commented:
Oh, right! So I bumped into something else than!
```
    fmt.Printf("Encrypted 1: %v\n", entity.PrivateKey.Encrypted)
    entity.PrivateKey.Encrypt([]byte("hello"), nil)
    fmt.Printf("Encrypted 2: %v\n", entity.PrivateKey.Encrypted)
    entity.PrivateKey.Decrypt([]byte("hello"))
    fmt.Printf("Encrypted 3: %v\n", entity.PrivateKey.Encrypted)
```
will print
```
Encrypted 1: false
Encrypted 2: true
Encrypted 3: true
```

But that's because `Decrypt` will bail out here:
```
	if pk.s2k == nil {
		return nil
	}
```

Would it be possible to set `s2k` properly so this workflow can work? It's not necessary, this is just something I made up and not an actual use case we would have right now.

kcwong-verseon committed:
Fix Decrypt() silently failed immediately after Encrypt()

* this is because the PrivateKey.s2k functor isn't set in
  Encrypt()
* add a test case for this
* improve previous test cases to check for the state of
  PrivateKey.Encrypted after Decrypt() and Encrypt() operations

dvusboy commented:
@zapu Thank you for the thorough review. I've addressed the issue you raised.

zapu commented on openpgp/packet/private_key.go:
```diff
@@ -187,6 +187,10 @@ func (pk *PrivateKey) Encrypt(passphrase []byte, config *Config) (err error) {
+	// No good way to set pk.s2k but to call s2k.Parse(),
+	// even though we have all the information here, but
+	// most of the functions needed are private to s2k.
+	pk.s2k, err = s2k.Parse(s2kBuf)
```
:+1: fine with me

zapu commented:
Thanks! Looks like my generate->encrypt->decrypt->serialize test program is now working correctly.

maxtaco commented:
Great work here @zapu  and @dvusboy!

maxtaco commented:
👍

dvusboy commented:
I'll rebase.

dvusboy commented:
Do you mind if I squash the commits?

zapu commented:
@maxtaco I think this looks ready

dvusboy commented:
Much appreciated.

maxtaco commented:
Thanks for your contributions!
